Add GUI module

Col-E opened the pull request:
## Preview

![image](https://github.com/Minecraft-Malware-Prevention-Alliance/concoction/assets/21371686/501035ff-a4c0-4c30-a692-e8262b3693fb)

![image](https://github.com/Minecraft-Malware-Prevention-Alliance/concoction/assets/21371686/7f4bf475-de45-4c4a-ab63-3c1d15f25ab4)


https://github.com/Minecraft-Malware-Prevention-Alliance/concoction/assets/21371686/d767fdf7-476f-425f-b2e4-ff3e3ad04910

Title says it all. Adding a common-use GUI.

## Pending contents

- [x] Progress reporting API in `lib` module _(Usage optional, but useful for UI feedback)_
- [x] Scanning UI implementation
    - [x] Core functionality
    - [x] Making it prettier
    - [x] Supporting exporting reports to file
    - [x] Supporting deleting files with detections
- [ ] Builder API for creating simplified UI's for specific malware campaigns _(pre-configured to target one thing, no user input really needed)_

Col-E committed:
Pull out common config to separate build files

Col-E committed:
Initial UI work for selecting inputs

Col-E commented on concoction-gui/src/main/java/info/mmpa/concoction/panel/TableModelPanel.java:
```diff
@@ -0,0 +1,189 @@
+		FileDialog dialog = new FileDialog(context.getFrame(), bundle.getString("input.select-files"));
+		dialog.setMultipleMode(true);
+		dialog.setMode(FileDialog.LOAD);
+		dialog.setDirectory("C:\\Code\\Java\\concoction\\concoction-cli\\build\\libs");
```
I forgot to remove my testing code here 😆

Col-E committed:
Allow API builder to swap out default path predicates for inputs/models

Col-E committed:
Outline of the scanning panel

Col-E committed:
Expose path as part of application per source model

Col-E committed:
Keep SourcePathElement allocations internal to ModelSource implementations

Since we use IdentityHashmap this will also make it easier for API consumers to get the same path reference we use since its baked into the model.

Col-E committed:
Fix wrong order in isAssignableFrom check

Col-E committed:
Add feedback API for scanning processes

Col-E committed:
Path access changes in model

Col-E committed:
Fix concoction api builder always attempting to run dynamic scan

Col-E committed:
Implement base scanning UI

Some cleaning up will come later

Col-E commented:
Huh, interesting failures. Works on my machine, and the failure on the builder test is really odd..

Col-E committed:
Fix equals / json decoding problems causing test failures

Col-E committed:
JDK 8 printstream internals caused this failure

Col-E committed:
Limit some log calls that can slow down UI behind debug flag

Col-E committed:
Wrap sink/feedback calls so that errors thrown by API consumers do not interfere with scan operations

Col-E committed:
Micro-optimization: Don't use stream() in multi-wrappers

Col-E committed:
Skip reading class debug info when instruction scanning

We don't use the data in instruction matching, so we can increase performance by skipping it.

Col-E committed:
Skip stringifying instructions we don't need to

Col-E committed:
Finish up scanning UI

Col-E commented:
GUI is basically done. Just need export support and management of actions on detected files.

Builder will then be done after that and re-use things from this GUI shown.

Col-E committed:
Create end panel which allows for saving of results and deleting of files with detections

xyzeva commented on settings.gradle:
```diff
@@ -1,3 +1,4 @@
-project(":concoction-lib").name = 'concoction'
\ No newline at end of file
+project(":concoction-lib").name = 'concoction'
+include 'concoction-gui'
```
Why is this here? Move this to the `include` above.

Col-E replied:
IntelliJ auto-add moment

xyzeva commented on concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java:
```diff
@@ -37,41 +37,33 @@ public DynamicMatchingModel(@Nonnull Map<String, DynamicMatchEntry> variants) {
 	 * @param archetype
 	 * 		Information about what the signature being matched.
-	 * @param sourcePath
+	 * @param path
```
Do we want to change this? Seems less descriptive to me.

Col-E replied:
Same as below comment

xyzeva commented on concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/DynamicMatchingModel.java:
```diff
@@ -37,41 +37,33 @@ public DynamicMatchingModel(@Nonnull Map<String, DynamicMatchEntry> variants) {
 	 */
 	public void matchOnExit(@Nonnull ResultsSink sink, @Nonnull DetectionArchetype archetype,
-							@Nonnull SourcePathElement sourcePath, @Nonnull CallStackFrame frame) {
+							@Nonnull MethodPathElement path, @Nonnull CallStackFrame frame) {
```
Do we want to change this? Seems less descriptive to me.

Col-E replied:
The type is changed. This should only ever consume a method path.

xyzeva replied:
Could we change it to `methodPath` then?

Col-E replied:
Done

Col-E commented on concoction-lib/src/main/java/info/mmpa/concoction/Concoction.java:
```diff
@@ -63,6 +68,39 @@ public Concoction withMaxInputDirectoryDepth(int inputDepth) {
 		return this;
 	}
 
+	/**
```
The changes were tied to things required by this GUI work, and there's some overlap so cherry-picking would be a pain since it'd require editing portions of commits

Col-E committed:
Some cleanup from PR review comments

Col-E committed:
Use custom patching solution to target Java 8 for GUI module with Java 11 dependency on Ikonli

It doesn't use any Java 11+ API's except in one case, which we rewrite to use 8 compliant API's.

Col-E commented:
Ok, Java 8 back-porting of Ikonli done. The GUI is not usable across the board when build with `:concoction-gui:shadowJar`

xyzeva commented on app-patching.gradle:
```diff
@@ -1,21 +1,95 @@
+        // an up-to-date version of ASM.
+        //
+        // I know, gradle is dumb.
+        classpath 'org.ow2.asm:asm-all:5.2'
```
Move this to `dependencies.gradle`, electric boogaloo 😢

xyzeva commented on settings.gradle:
```diff
@@ -1,3 +1,15 @@
+    // Used for patching in the GUI module to use a Java 11 dependency on Java 8.
+    // See 'app-patching.gradle' for more details.
+    dependencies {
+        classpath 'org.ow2.asm:asm:9.5'
```
Move to `dependencies.gradle`, tokyo drift 😿

xyzeva reviewed (approved):
LGTM
